harness/determined/pytorch/_pytorch_trial.py

```diff
@@ -271,75 +498,80 @@ def cleanup_iterator() -> None:
+                    for op in self.core_context.searcher.operations():
+                        self.searcher_op = op
+
+                        searcher_unit = self.core_context.searcher.get_configured_units()
+
+                        # We only support batches for now
+                        if searcher_unit == core.Unit.EPOCHS:
+                            train_unit = Batch(self.as_batches(batches=None, records=None, epochs=op.length))
+                        elif searcher_unit == core.Unit.BATCHES:
+                            train_unit = Batch(op.length)
+                        elif searcher_unit == core.Unit.RECORDS:
+                            train_unit = Batch(self.as_batches(batches=None, records=op.length, epochs=None))
+                        else:
+                            raise ValueError(f"Train unit {searcher_unit} not supported.")
 
-                else:
-                    raise AssertionError("Unexpected workload: {}".format(w.kind))
+                        min_checkpoint_period = self.as_batches(**min_checkpoint_period)
+                        min_validation_period = self.as_batches(**min_validation_period)
+                        if min_checkpoint_period < 1:
+                            min_checkpoint_period = sys.maxsize
+
+                        if min_validation_period < 1:
+                            min_validation_period = sys.maxsize
+
+                        self._train(train_unit=train_unit,
+                                    min_validation_period=min_validation_period,
+                                    min_checkpoint_period=min_checkpoint_period)
+
+                        logging.debug(f"Training finished after {train_unit.value} steps.")
+
+                        self.training_metrics = self.context.distributed.broadcast(self.training_metrics)
+                        for callback in self.callbacks.values():
+                            callback.on_training_workload_end(
+                                avg_metrics=self.training_metrics["avg_metrics"],
+                                batch_metrics=self.training_metrics["batch_metrics"],
+                            )
+
+                        # Report metrics after searcher complete if there are remaining unreported
+                        if self.training_metrics:
+                            self.report_training_metrics(self.steps_completed, self.training_metrics)
+
+                        # Validate and report validation metrics if last validation step is not current step
+                        if self.save_state.last_val != self.steps_completed:
+                            val_metrics = self._validate()
+                            self.report_validation_metrics(self.steps_completed, val_metrics)
+
+                        # Checkpoint
+                        self.checkpoint()
+                except ShouldExit as ex:
+                    print("exit requested, saving checkpoint")
+                    self.checkpoint()
```

Comment:
should you check if `ex.skip_exit_checkpoint` is set?  (full disclosure: I don't remember what that flag is actually for)

Reply:
done